File: src/main/java/org/apache/fineract/client/services/LoanCollateralApi.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.fineract.client.services;

import org.apache.fineract.client.CollectionFormats.*;

import rx.Observable;
import retrofit2.http.*;

import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import okhttp3.MultipartBody;

import org.apache.fineract.client.models.DeleteLoansLoanIdCollateralsCollateralIdResponse;
import org.apache.fineract.client.models.GetLoansLoanIdCollateralsResponse;
import org.apache.fineract.client.models.GetLoansLoanIdCollateralsTemplateResponse;
import org.apache.fineract.client.models.PostLoansLoanIdCollateralsRequest;
import org.apache.fineract.client.models.PostLoansLoanIdCollateralsResponse;
import org.apache.fineract.client.models.PutLoansLoanIdCollateralsCollateralIdResponse;
import org.apache.fineract.client.models.PutLoansLoandIdCollateralsCollateralIdRequest;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public interface LoanCollateralApi {
  /**
   * Create a Collateral
   * Note: Currently, Collaterals may be added only before a Loan is approved
   * @param loanId loanId (required)
   * @param postLoansLoanIdCollateralsRequest  (required)
   * @return Observable&lt;PostLoansLoanIdCollateralsResponse&gt;
   */
  @Headers({
    "Content-Type:application/json"
  })
  @POST("loans/{loanId}/collaterals")
  Observable<PostLoansLoanIdCollateralsResponse> createCollateral(
    @retrofit2.http.Path("loanId") Long loanId, @retrofit2.http.Body PostLoansLoanIdCollateralsRequest postLoansLoanIdCollateralsRequest
  );

  /**
   * Remove a Collateral
   * Note: A collateral can only be removed from Loans that are not yet approved.
   * @param loanId loanId (required)
   * @param collateralId collateralId (required)
   * @return Observable&lt;DeleteLoansLoanIdCollateralsCollateralIdResponse&gt;
   */
  @DELETE("loans/{loanId}/collaterals/{collateralId}")
  Observable<DeleteLoansLoanIdCollateralsCollateralIdResponse> deleteCollateral(
    @retrofit2.http.Path("loanId") Long loanId, @retrofit2.http.Path("collateralId") Long collateralId
  );

  /**
   * Retrieve Collateral Details Template
   * This is a convenience resource. It can be useful when building maintenance user interface screens for client applications. The template data returned consists of any or all of:  Field Defaults Allowed Value Lists Example Request:  loans/1/collaterals/template
   * @param loanId loanId (required)
   * @return Observable&lt;GetLoansLoanIdCollateralsTemplateResponse&gt;
   */
  @GET("loans/{loanId}/collaterals/template")
  Observable<GetLoansLoanIdCollateralsTemplateResponse> newCollateralTemplate(
    @retrofit2.http.Path("loanId") Long loanId
  );

  /**
   * List Loan Collaterals
   * Example Requests:  loans/1/collaterals   loans/1/collaterals?fields&#x3D;value,description
   * @param loanId loanId (required)
   * @return Observable&lt;List&lt;GetLoansLoanIdCollateralsResponse&gt;&gt;
   */
  @GET("loans/{loanId}/collaterals")
  Observable<List<GetLoansLoanIdCollateralsResponse>> retrieveCollateralDetails(
    @retrofit2.http.Path("loanId") Long loanId
  );

  /**
   * Retrieve a Collateral
   * Example Requests:  /loans/1/collaterals/1   /loans/1/collaterals/1?fields&#x3D;description,description
   * @param loanId loanId (required)
   * @param collateralId collateralId (required)
   * @return Observable&lt;GetLoansLoanIdCollateralsResponse&gt;
   */
  @GET("loans/{loanId}/collaterals/{collateralId}")
  Observable<GetLoansLoanIdCollateralsResponse> retrieveCollateralDetails1(
    @retrofit2.http.Path("loanId") Long loanId, @retrofit2.http.Path("collateralId") Long collateralId
  );

  /**
   * Update a Collateral
   * 
   * @param loanId loanId (required)
   * @param collateralId collateralId (required)
   * @param putLoansLoandIdCollateralsCollateralIdRequest  (required)
   * @return Observable&lt;PutLoansLoanIdCollateralsCollateralIdResponse&gt;
   */
  @Headers({
    "Content-Type:application/json"
  })
  @PUT("loans/{loanId}/collaterals/{collateralId}")
  Observable<PutLoansLoanIdCollateralsCollateralIdResponse> updateCollateral(
    @retrofit2.http.Path("loanId") Long loanId, @retrofit2.http.Path("collateralId") Long collateralId, @retrofit2.http.Body PutLoansLoandIdCollateralsCollateralIdRequest putLoansLoandIdCollateralsCollateralIdRequest
  );

}

File: src/main/kotlin/org/openapitools/client/apis/LoanCollateralApi.kt
package org.openapitools.client.apis

import org.openapitools.client.models.DeleteLoansLoanIdCollateralsCollateralIdResponse
import org.openapitools.client.models.GetLoansLoanIdCollateralsResponse
import org.openapitools.client.models.GetLoansLoanIdCollateralsTemplateResponse
import org.openapitools.client.models.PostLoansLoanIdCollateralsRequest
import org.openapitools.client.models.PostLoansLoanIdCollateralsResponse
import org.openapitools.client.models.PutLoansLoanIdCollateralsCollateralIdResponse
import org.openapitools.client.models.PutLoansLoandIdCollateralsCollateralIdRequest
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path

interface LoanCollateralApi {
    /**
     * Create a Collateral
     * Note: Currently, Collaterals may be added only before a Loan is approved
     * Responses:
     *  - 200: OK
     *
     * @param loanId loanId
     * @param postLoansLoanIdCollateralsRequest
     * @return [PostLoansLoanIdCollateralsResponse]
     */
    @POST("v1/loans/{loanId}/collaterals")
    suspend fun createCollateral(
        @Path("loanId") loanId: Long,
        @Body postLoansLoanIdCollateralsRequest: PostLoansLoanIdCollateralsRequest
    ): PostLoansLoanIdCollateralsResponse

    /**
     * Remove a Collateral
     * Note: A collateral can only be removed from Loans that are not yet approved.
     * Responses:
     *  - 200: OK
     *
     * @param loanId loanId
     * @param collateralId collateralId
     * @return [DeleteLoansLoanIdCollateralsCollateralIdResponse]
     */
    @DELETE("v1/loans/{loanId}/collaterals/{collateralId}")
    suspend fun deleteCollateral(
        @Path("loanId") loanId: Long,
        @Path("collateralId") collateralId: Long
    ): DeleteLoansLoanIdCollateralsCollateralIdResponse

    /**
     * Retrieve Collateral Details Template
     * This is a convenience resource. It can be useful when building maintenance user interface screens for client applications. The template data returned consists of any or all of:  Field Defaults Allowed Value Lists Example Request:  loans/1/collaterals/template
     * Responses:
     *  - 200: OK
     *
     * @param loanId loanId
     * @return [GetLoansLoanIdCollateralsTemplateResponse]
     */
    @GET("v1/loans/{loanId}/collaterals/template")
    suspend fun newCollateralTemplate(@Path("loanId") loanId: Long): GetLoansLoanIdCollateralsTemplateResponse

    /**
     * List Loan Collaterals
     * Example Requests:  loans/1/collaterals   loans/1/collaterals?fields&#x3D;value,description
     * Responses:
     *  - 200: OK
     *
     * @param loanId loanId
     * @return [kotlin.collections.List<GetLoansLoanIdCollateralsResponse]
     */
    @GET("v1/loans/{loanId}/collaterals")
    suspend fun retrieveCollateralDetails(@Path("loanId") loanId: Long): List<GetLoansLoanIdCollateralsResponse>

    /**
     * Retrieve a Collateral
     * Example Requests:  /loans/1/collaterals/1   /loans/1/collaterals/1?fields&#x3D;description,description
     * Responses:
     *  - 200: OK
     *
     * @param loanId loanId
     * @param collateralId collateralId
     * @return [GetLoansLoanIdCollateralsResponse]
     */
    @GET("v1/loans/{loanId}/collaterals/{collateralId}")
    suspend fun retrieveCollateralDetails1(
        @Path("loanId") loanId: Long,
        @Path("collateralId") collateralId: Long
    ): GetLoansLoanIdCollateralsResponse

    /**
     * Update a Collateral
     *
     * Responses:
     *  - 200: OK
     *
     * @param loanId loanId
     * @param collateralId collateralId
     * @param putLoansLoandIdCollateralsCollateralIdRequest
     * @return [PutLoansLoanIdCollateralsCollateralIdResponse]
     */
    @PUT("v1/loans/{loanId}/collaterals/{collateralId}")
    suspend fun updateCollateral(
        @Path("loanId") loanId: Long,
        @Path("collateralId") collateralId: Long,
        @Body putLoansLoandIdCollateralsCollateralIdRequest: PutLoansLoandIdCollateralsCollateralIdRequest
    ): PutLoansLoanIdCollateralsCollateralIdResponse

}
